src/Microsoft.AspNet.Security.OAuth/OAuthBearerAuthenticationHandler.cs

```diff
@@ -24,87 +35,167 @@ protected override AuthenticationTicket AuthenticateCore()
+						requestTokenContext.Token = authorization.Substring("Bearer ".Length).Trim();
                     }
                 }
 
-                // Give application opportunity to find from a different location, adjust, or reject token
-                var requestTokenContext = new OAuthRequestTokenContext(Context, requestToken);
-                await Options.Notifications.RequestToken(requestTokenContext);
-
                 // If no token found, no further work possible
                 if (string.IsNullOrEmpty(requestTokenContext.Token))
                 {
                     return null;
                 }
 
-                // Call provider to process the token into data
-                var tokenReceiveContext = new AuthenticationTokenReceiveContext(
-                    Context,
-                    Options.AccessTokenFormat,
-                    requestTokenContext.Token);
-
-                await Options.AccessTokenProvider.ReceiveAsync(tokenReceiveContext);
-                if (tokenReceiveContext.Ticket == null)
+				// notify user token was received
+				var securityTokenReceivedNotification = 
+					new SecurityTokenReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>(Context, Options)
+					{
+						ProtocolMessage = requestTokenContext,
+					};
+
+				await Options.Notifications.SecurityTokenReceived(securityTokenReceivedNotification);
+				if (securityTokenReceivedNotification.HandledResponse)
+				{
+					return securityTokenReceivedNotification.AuthenticationTicket;
+				}
+
+				if (securityTokenReceivedNotification.Skipped)
+				{
+					return null;
+				}
+
+				if (_configuration == null)
                 {
-                    tokenReceiveContext.DeserializeTicket(tokenReceiveContext.Token);
+                    _configuration = await Options.ConfigurationManager.GetConfigurationAsync(Context.RequestAborted);
                 }
 
-                AuthenticationTicket ticket = tokenReceiveContext.Ticket;
-                if (ticket == null)
-                {
-                    _logger.WriteWarning("invalid bearer token received");
-                    return null;
-                }
+				var validationParameters = Options.TokenValidationParameters.Clone();
+				if (validationParameters.ValidIssuer == null && !string.IsNullOrWhiteSpace(_configuration.Issuer))
+				{
+					validationParameters.ValidIssuer = _configuration.Issuer;
+				}
+				else
+				{
+					IEnumerable<string> issuers = new[] { _configuration.Issuer };
+					validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
+				}
+
+				validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? _configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(_configuration.SigningKeys));
+				SecurityToken validatedToken;
+				foreach (var validator in Options.SecurityTokenValidators)
```

Comment:
foreach loop made me think we are creating a separate principal for every validator. Instead will it be better to say 

```
var validator = Options.SecurityTokenValidators.Where(v => v.CanReadToken(requestTokenContext.Token)).FirstOrDefault();
```

And do the validation and return from here.